WIP Immutable dto support with config

phoenix128 opened the pull request:
Mutable/Immutable DTO support through declarative schema.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)


### Resolved issues:
1. [x] resolves magento/magento2#30037: WIP Immutable dto support with config

phoenix128 committed:
Immutable data object helper to manipulate immutable objects in Magento

phoenix128 committed:
Less ugly test

phoenix128 committed:
Codestyling fixes

phoenix128 committed:
Data transport helper and test coverage

phoenix128 committed:
Reworking dto processors

phoenix128 committed:
WIP ServiceInputProcessor rework

phoenix128 committed:
WIP: Custom attributes support

phoenix128 committed:
MOdified orphan handling

phoenix128 committed:
FIX method mapping

phoenix128 committed:
PSR-7 like mutators

phoenix128 committed:
FIX array handling

phoenix128 committed:
FIX extension attributes for dto processor

phoenix128 committed:
Extension attribute mutable injector

phoenix128 committed:
FIX type array

phoenix128 committed:
FIX type casting

phoenix128 committed:
FIX extension attributes

phoenix128 committed:
Backward compatibility mode

phoenix128 committed:
Test immutable processor

phoenix128 committed:
Test with immutable extension attributes interface

phoenix128 committed:
WIP DTO autogenerator

phoenix128 committed:
Test coverage and data processing

phoenix128 committed:
Full PSR-7 support for immutables

phoenix128 committed:
PSR7 test coverage

phoenix128 committed:
DataObjectHelper integration with DtoSupport

phoenix128 committed:
Hydrator strategy

phoenix128 committed:
Rework for Service Input Processor backward compatiblity
- Rework ServiceInput Processor using createFromArray method of DtoProcessor
- ServiceInputPreocessorTest moved to integration tests
- Support for custom attributes in DtoProcessor
- Additional test coverage

phoenix128 committed:
Ejected Hydration strategy and reflection + test coverage

YevSent commented on lib/internal/Magento/Framework/Api/DataObjectHelper.php:
```diff
@@ -45,28 +54,46 @@ class DataObjectHelper
+        ExtensionAttributesFactory $extensionFactory,
+        JoinProcessorInterface $joinProcessor,
+        MethodsMap $methodsMapProcessor,
+        DtoProcessor $dtoProcessor = null,
```
New dependencies can be added as required arguments.

YevSent commented on lib/internal/Magento/Framework/Api/ExtensionAttribute/InjectorProcessorInterface.php:
```diff
@@ -0,0 +1,25 @@
+/**
+ * Extension attributes injectors processor interface
+ *
+ * @api
```
@api annotation can be added in major release.

YevSent commented on lib/internal/Magento/Framework/Api/ExtensionAttributesFactory.php:
```diff
@@ -27,48 +38,66 @@ class ExtensionAttributesFactory
+        TypeProcessor $typeProcessor = null
+    ) {
         $this->objectManager = $objectManager;
+        $this->typeProcessor = $typeProcessor ?: ObjectManager::getInstance()
```
The argument can be added as required.

YevSent commented on lib/internal/Magento/Framework/Dto/Code/Generator/Mutator.php:
```diff
@@ -0,0 +1,91 @@
+    /**
+     * Entity type
+     */
+    const ENTITY_TYPE = 'mutator';
```
Please, add scope visibility for the constant.

YevSent commented on lib/internal/Magento/Framework/Dto/Code/GetMutatorSourceCode.php:
```diff
@@ -0,0 +1,230 @@
+    }
+
+    /**
+     * @param string $resultClassName
```
Probably the static test will fail on missed short method's description.

YevSent commented on lib/internal/Magento/Framework/Dto/Config/Reader.php:
```diff
@@ -0,0 +1,57 @@
+use Magento\Framework\Config\Reader\Filesystem;
+use Magento\Framework\Config\ValidationStateInterface;
+
+class Reader extends Filesystem
```
Probably the static test will fail on the missed class description.

YevSent commented on lib/internal/Magento/Framework/Dto/Config/SchemaLocator.php:
```diff
@@ -0,0 +1,50 @@
+    /**
+     * @var UrnResolver
+     */
+    protected $urnResolver;
```
Why the visibility scope is `protected`?

YevSent commented on lib/internal/Magento/Framework/Dto/DtoProcessor/DtoReflection.php:
```diff
@@ -0,0 +1,165 @@
+     */
+    public function isExtensibleObject(string $className): bool
+    {
+        // TODO: Add a cache layer here
```
`@TODO` annotations should be avoided.

YevSent commented on lib/internal/Magento/Framework/Dto/DtoProcessor/GetHydrationStrategy.php:
```diff
@@ -0,0 +1,161 @@
+    /**
+     * Strategy for setter hydration
+     */
+    public const HYDRATOR_STRATEGY_SETTER = 'setter';
```
Should be these constants as public?

phoenix128 replied:
They are required by other classes like `\Magento\Framework\Dto\DtoProcessor`

YevSent commented on lib/internal/Magento/Framework/Dto/DtoProcessor/GetHydrationStrategy.php:
```diff
@@ -0,0 +1,161 @@
+     * @return array
+     * @throws ReflectionException
+     */
+    public function execute(string $className, array $data): array
```
I would like to propose splitting up this method into multiple for better readability.

YevSent commented on lib/internal/Magento/Framework/Dto/Test/Unit/Code/Generator/DtoGeneratorTest.php:
```diff
@@ -0,0 +1,170 @@
+    private $getDtoSourceCode;
+
+    /**
+     * @var DtoConfig|PHPUnit_Framework_MockObject_MockObject
```
`PHPUnit\Framework\MockObject\MockObject` should be used instead.

YevSent commented on dev/tests/integration/testsuite/Magento/Framework/Reflection/Mock/TestExtensionAttributes.php:
```diff
@@ -0,0 +1,48 @@
+use Magento\Framework\Api\AbstractSimpleObject;
+use Magento\Framework\Api\ExtensionAttributesInterface;
+
+class TestExtensionAttributes  extends AbstractSimpleObject implements ExtensionAttributesInterface
```
Extra space

phoenix128 committed:
Merge branch '2.3-develop' into immutable-dto-support-with-config

phoenix128 committed:
New dependencies can be added as required arguments

phoenix128 committed:
@api annotation can be added in major release.

phoenix128 committed:
The argument can be added as required.

phoenix128 committed:
Missing constant scope

phoenix128 committed:
Missing short doc

phoenix128 committed:
Wrong mock class

phoenix128 committed:
FIX CS: Extra space

phoenix128 committed:
Merge branch '2.3-develop' into immutable-dto-support-with-config

phoenix128 committed:
FIX unit tests

phoenix128 committed:
FIX broken ServiceInputProcessor after merge

phoenix128 committed:
FIX race condition in code generation

phoenix128 committed:
FIX time coupling issue on code generation

phoenix128 committed:
Add exception with orphan parameters

phoenix128 committed:
FIX CS after test fix

phoenix128 committed:
FIX tests

phoenix128 committed:
FIX missing classes for unit tests

phoenix128 committed:
FIX test

phoenix128 committed:
FIX assay management in extension attributes and broken tests

phoenix128 committed:
FIX simple types array in extension attributes

phoenix128 committed:
FIX error message for custom attributes

phoenix128 committed:
Remove unnecessary check on return type

phoenix128 committed:
FIX extension attributes interface name error

phoenix128 committed:
Reverted commit for unneded check reintroduced by mistake

phoenix128 committed:
Skip extension attributes creation if no data exists
+ FIXES issues with non standard extension attributes like AttributeExtensionInterface

phoenix128 committed:
FIX creation of extnesion attributes object

phoenix128 committed:
FIX test with no type casting

phoenix128 committed:
No casting types in service input processor

phoenix128 committed:
FIX non array value for backward compatibilty with old ServiceInputProcesor behaviour

phoenix128 committed:
FIX null values in object creation while using setters strategy

phoenix128 committed:
FIX wrong sorting order definition in test

phoenix128 committed:
Check data format while creating objects

phoenix128 committed:
FIX typing issues

phoenix128 committed:
FIX type casting issues

phoenix128 committed:
FIX error handling

phoenix128 committed:
Merge branch '2.3-develop' into immutable-dto-support-with-config

phoenix128 committed:
FIX CS

phoenix128 committed:
Config change for isolation support

phoenix128 committed:
FIX CS

phoenix128 committed:
WIP rework config

phoenix128 committed:
Auto preference for DTO interfaces

phoenix128 committed:
Test coverage for extension attributes direct injection

phoenix128 committed:
Test coverage for extension attributes injection + FIX preferenced name on injector definition

phoenix128 committed:
FIX broken unit test

phoenix128 committed:
FIX CS

phoenix128 committed:
FIX wrong variable scope

phoenix128 committed:
DTO projection

phoenix128 committed:
Rework of straight mapping and pre/post processors

phoenix128 committed:
FIX create via dto processor

phoenix128 committed:
Merge branch '2.3-develop' into immutable-dto-support-with-config

phoenix128 committed:
FIX dto.xsd declaration

phoenix128 committed:
DTO factory code generation

sidolov commented:
@magento create issue

Den4ik reviewed (changes requested):
@phoenix128 Thanks for work. Is it still actual and WIP?
If yes, please resolve merge conflicts

ishakhsuvarov commented:
Hi @phoenix128 

I am closing this PR for now as it seems like it is outdated and still needs a lot of work.

As usual, please feel free to reopen if you wish to continue.

Cheers!
